### PR TITLE
[1.18.x] Added a hook to activate spyglass-style scoping

### PR DESCRIPTION
**Pull request by Da-Technomancer**
This PR is a previous PR (https://github.com/MinecraftForge/MinecraftForge/pull/8171), rebased onto 1.18.

Added a hook to trigger the 'scoping' effect the spyglass uses (zoom in, visual overlay, adjusted movement).

Mods can register a predicate which will be tested on the client side to see whether the scoping effect is being externally triggered. If any registered predicate returns true, scoping will be active.

Also wrote a test mod.

### CODE REVIEW

**Comment by Shadows-of-Fire** [comment]
Looks fine to me, would go ahead and assign but the CI is failing and I can't check if it compiles manually right now.

**Comment by Da-Technomancer** [comment]
CI is saying it passes now. I didn't make any changes.

**Comment by Shadows-of-Fire** [comment]
Yeah it just had to be restarted, figured that was the case.

File: src/main/java/net/minecraftforge/common/SpyglassHooks.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+package net.minecraftforge.common;
+
+import net.minecraft.world.entity.player.Player;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class SpyglassHooks
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/SpyglassHooks.java]: This doesn't need to be its own class, you can use `ForgeHooks`.

File: src/main/java/net/minecraftforge/common/SpyglassHooks.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * Registers a predicate which will be checked for whether the player is considered to be 'scoping' with a spyglass.
+     * The predicate will only be tested on the client.
+     * Only one registered predicate needs to be true for scoping to occur.
+     * <strong>The passed predicate will be tested very frequently (typically every frame).</strong>
+     *
+     * @param spyglassCondition A predicate which will be checked on the client side.
+     */
+    public static synchronized void registerSpyglassCondition(Predicate<Player> spyglassCondition)
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/SpyglassHooks.java]: This is not truly thread safe.
Please instead fire an event at startup to collect all conditions and then store them in an immutable and therefor threadsafe list.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
Please update the license headers to 2022 as well and resolve merge conflicts.

File: src/main/java/net/minecraftforge/common/SpyglassHooks.java (hunk 2 above)
Comment (Da-Technomancer) [reply src/main/java/net/minecraftforge/common/SpyglassHooks.java]: Could you clarify why you consider it thread-unsafe? I must be missing something, but I don't see it.

Comment (diesieben07) [reply src/main/java/net/minecraftforge/common/SpyglassHooks.java]: `SPYGLASS_PREDICATES` is accessed from two methods. One of them is synchronized, one of them is not.
However that is irrelevant here, the more important change request is to add a dedicated collection event.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
I don't think these should be a full blown registry. Why does it need to be?

**Comment by Da-Technomancer** [comment]
> I don't think these should be a full blown registry. Why does it need to be?

It had seemed cleaner to use the tried-and-tested registry model than to create a new system for registration at mod startup. I'm open to suggestions.

**Comment by diesieben07** [comment]
Fire an event. That event has an internal list of predicates and a method `addPredicate`. After the event is fired, bake the list into an immutable one that is then stored and used later. This avoids threading issues and the need for synchronization, it improves performance, etc.

**Comment by Da-Technomancer** [comment]
After several revisions, I'm calling it quits.
I was not able to find a way to fire an event only as needed which was both robust and cleanly implemented; I'm sure a way exists, but someone more familiar with the inner workings of Forge would need to find it. 

As a workaround, for anyone who found this PR because they wanted to add their version of a spyglass feature, here's how I suggest doing it without relying on this PR. It's only a rough approximation of what this PR is meant to achieve.
- Achieve the 'zoom' effect using an FOV modifier. There is an existing forge hook for FOV; see 'ForgeHooksClient::getFieldOfView' and set the FOV to 0.1 when zooming is desired. 
- Achieve the spyglass overlay effect using an event handler for the game rendering, and create your own overlay render.
- Achieve the sound by playing the sound (see the vanilla spyglass item class)

**Pull request closed without merging**